### PR TITLE
fix(history): record the modes of directories that contain tracked files

### PR DESCRIPTION
**Pull request by jdx**
https://entire.io/gh/jdx/mise/trails/58

Tracking an individual file inside a private directory did not record the directory's mode, so a fresh machine recreated the directory world-readable. With this configuration and `~/.claude` at `0700`:

```toml
[dotfiles]
"~/.claude/settings.json" = { mode = "track" }
```

the history manifest recorded nothing about `~/.claude`, and `mise bootstrap --from-git` on a new machine created it with the default `0755` via `create_dir_all`:

```json
// before
"permissions": {}
```

Directory modes were only captured for directories that were themselves `[dotfiles]` entries (and their subdirectories). Reported in #13410 (item 3).

## What changes

A checkpoint now records the non-default (not `0755`) mode of every directory strictly between a tracked path and the root it is mapped under (home or the mise configuration directory). The same rule applies to file and directory entries, so a tracked `~/.config/nvim` inside a `0700` `~/.config` records `~/.config` too.

```json
// after
"permissions": {
  "home/.claude": 448
}
```

A containing directory is one filesystem object, so it gets one record under the variant-less path whatever streams the files inside it belong to: `~/.claude/settings.json` without variants and `~/.claude/notes` with `variants = [{ os = "linux" }]` share the single `home/.claude` record, and a save from either stream replaces it. The record applies even when every file inside the directory is platform-specific, so the tree only holds it under a stream such as `home@linux/.claude`. A directory that is itself an enrolled entry keeps its own per-stream record (`home@linux/.claude`), as before; when the directory is enrolled here, that stream's record decides its mode and its absence means the default, so another platform's containing record for the same directory does not apply (an enrolled stream the incoming manifest dropped means a machine with that stream saved the directory at the default). A stream marker lives only in the first path component, so a directory whose own name contains `@` is an ordinary directory.

On pull and on a fresh `bootstrap --from-git`, the directory is created (or corrected) with the recorded mode before its files are written. When the directory goes back to `0755` at home, the next save drops the record and the next pull on the other machine restores the default. A machine whose directory already has the incoming mode pulls without saving first, even if its own manifest predates this change and records nothing for it; only a live mode that differs from the incoming one is checked against the local record. That check has one purpose: never overwrite a live mode the user has not saved. The baseline it compares the live mode with is the mode this planner would have applied from the saved manifest: the same computation that decides the incoming mode (the enrolled stream's record covering the directory, absent meaning the default; else the containing record, absent meaning the default), run against the saved manifest with this machine's variant selection. A directory the saved enrollment does not cover has no baseline, and a directory that does not exist live has nothing to guard. Because what a pull applies and what the next pull assumes are one definition, a mode this planner set is never reported back as an unsaved change, tree presence plays no part in the baseline, and a nested enrollment for another platform cannot hide the enclosing enrollment that governs the directory here. The same governing key decides whether a directory step exists at all, in three ways: a file or link at that key in the incoming tree means the path is never turned into a directory because another stream holds one there; a directory at that key means present; and nothing at that key (every file inside lives in a nested variant stream) falls through to the containing rule, where the directory is present when some stream of a path inside it holds it as a directory and none holds something else there. A checkpoint records a directory's mode only from the record under its governing key, so another platform's containing record does not become this machine's rollback target when its own stream records the default. The checkpoint record's `tree.modes` includes these directories too, so `history rollback` recreates them privately and an unchanged tree still deduplicates on automatic captures.

A selective save of one child of a manual (`autosave = false`) directory now records the live mode of every directory the save observed (the manual directory itself, its subdirectories, its parents) while unnamed sibling files keep their saved modes; previously the directory's own mode was stripped as an unnamed sibling and only put back from the previous checkpoint, which could not distinguish an unchanged mode from a stale one.

## Deliberately still not recorded

- Home itself and the mise configuration directory itself. Enrolling a file must not capture metadata about your home or a custom configuration root, and a fresh machine's home keeps its own mode.
- Directories whose mode is the default `0755`, as before.
- For a path outside every root (not under home or the configuration directory), only the explicitly enrolled directory and its subdirectories, as before.
- A per-stream record for a containing directory (`home@linux/.claude`): the manifest validator rejects it, and a sibling, the root, or a directory containing no enrolled path is rejected as before.

## Known follow-up (pre-existing)

Directory permission steps run before the file steps of the same pull, both on `main` (for enrolled directories) and here (now also for containing directories). A recorded non-writable mode such as `0555` is therefore applied before files are written into that directory, and the writes fail with permission denied. That ordering is unchanged by this PR; a contained fix (apply `desired | 0o700` before the file steps and `desired` after, with the temporary mode covered by rollback and concurrent-change validation) is left for a follow-up.

## Known limitation (not changed here)

A containing directory has one variant-less record. Two machines that keep the same shared parent at different modes, say `~/.claude` at `0700` on Linux and `0750` on macOS, overwrite each other's `home/.claude` record on each save, and each pull applies the other's mode. Nothing is lost or aborted (the unsaved-mode guard still reports a change made after the local machine's last save), but a per-platform mode for a shared containing directory cannot be expressed. Two directions are possible and the choice is the maintainer's: record a containing directory per stream when the files under it are platform-specific, or keep the single record and document last-writer-wins.

## Validation

- Unit: `cargo test --bin mise history::` (186 passed). New cases cover: a file entry under a `0700` parent records the parent and never home, and the configuration root is its own boundary (`checkpoint.rs`); capture, replacement on return to default (including a directory named with `@`), one record across streams, and an inactive directory enrollment for another platform not hiding the parent record (`manifest.rs`); eligibility of containing directories, one step per directory when an enrolled stream and a containing record name the same directory (with the enrolled stream authoritative even when only the local head records it), a parent present only under a variant stream being planned on a fresh machine and still guarding an unsaved local chmod, a superseded containing key still guarding an unsaved local chmod, a dropped enrolled record read as the default rather than another platform's containing record, a directory newly enrolled in a stream where only a child was enrolled keeping its containing baseline, the inverse (a variant enrollment reduced to a containing record) keeping the record under the selected variant, a new private subdirectory inside an enrolled directory being created rather than guarded, a nested enrollment for another platform not hiding the enclosing enrollment's baseline, adoption and the next pull agreeing on the baseline (a containing change after adopting is planned or needs nothing, never a conflict), a path that is a file in its governing stream getting no directory step even though another stream holds a directory there, a covered directory that exists only in an inner variant stream still getting its recorded mode on a fresh machine, and a matching parent with no local record not being treated as unsaved (`sync/directories.rs`); parent modes reaching the checkpoint record with deduplication, a checkpoint taking a directory's mode only from its governing stream (`shadow.rs`), and a selective save of a manual directory's child neither carrying a stale parent mode nor dropping the directory's unchanged mode (`checkpoint.rs`). Each fix was checked by reverting it and watching its test fail.
- E2E: `e2e/cli/test_dotfiles_directory_permissions` now tracks `~/.claude/settings.json` under a `0700` `~/.claude` and a `~/.variant-private/settings.json` declared with `--os` under a `0700` `~/.variant-private`; asserts the manifest records `home/.claude` and `home/.variant-private` as `448`, nothing for `home` or the per-stream path, and the file under `home@<os>/`; adopts on a second HOME and asserts both directories come back `0700` while that HOME stays `0755`; then chmods `~/.claude` back to `0755`, saves, and asserts the pull on the second machine follows. `test_dotfiles_portable_permissions`, `test_dotfiles_rollback_types`, `test_dotfiles_history_policies`, `test_dotfiles_capture_omissions`, and `test_dotfiles_files` pass as well.
- `cargo clippy --bin mise --all-features --all-targets -- -D warnings` and `mise run lint-fix` are clean.

*AI-assisted — Tool: Claude Code; model: anthropic/claude-fable-5-1; version: unavailable.*

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes Unix directory permission capture, sync, and pull conflict detection for dotfiles history—security-sensitive—but behavior is heavily specified and covered by new unit and E2E tests.
> 
> **Overview**
> **Dotfiles history now persists and restores non-default modes for parent directories** of tracked files (e.g. `0700` on `~/.claude` when only `settings.json` is enrolled), so bootstrap/pull no longer recreate those paths as world-readable `0755`.
> 
> Capture walks ancestors up to—but not including—home or the mise config root via `mode_ancestors` / updated `file_modes`, writes a **single variant-less** manifest key for shared containing dirs (with validation in `manifest.rs`), and keeps live observed modes on selective manual saves instead of stale carried modes. **`governing_key` / `mode_from`** unify how desired modes are chosen across checkpoint metadata, rollback (`shadow.rs`), and the **rewritten directory pull planner** (`sync/directories.rs`): enrolled-stream records win over cross-platform containing records, tree presence avoids turning files into directories, and unsaved local chmods are gated against the same baseline the last saved manifest would have applied.
> 
> Docs (`history.md`) and `e2e/cli/test_dotfiles_directory_permissions` cover `~/.claude` and OS-variant parents; extensive unit tests were added for variant/enrollment edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 309250b5125e6d99528b3241f37d07cf82701972. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/system/history/sync/directories.rs [hunk 1]
```diff
@@ -54,7 +74,7 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
         .chain(saved.permissions.keys())
         .collect();
     for portable in paths {
-        if !super::run::eligible(&roots, tracked, portable)
+        if !eligible(&roots, tracked, portable)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed: a machine whose parent already sat at the incoming mode bailed because its pre-upgrade manifest recorded nothing for it (read as 0755). Fixed in f7ee0087b: a live mode equal to the incoming one is never checked against the local record; only a differing live mode is. Covered by the new unit test matching_parent_without_a_local_record_is_not_unsaved.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/checkpoint.rs [hunk 2]
```diff
@@ -736,7 +750,11 @@ impl Store {
                     .tree
                     .modes
                     .into_iter()
-                    .filter(|(path, _)| entries.iter().any(|entry| under_entry(path, entry)))
+                    .filter(|(path, _)| {
+                        entries
+                            .iter()
+                            .any(|entry| under_entry(path, entry) || under_entry(entry, path))
+                    })
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs:762]: Confirmed, though it was masked: read_meta dropped containing directories from the record tree.modes, so saved_modes never returned them (which also broke rollback and dedup for these directories). f7ee0087b fixes read_meta and applies the same observed-directory guard to the selective promotion path; the new selective_save_does_not_restore_a_stale_parent_mode test fails without the guard and passes with it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/manifest.rs [hunk 3]
```diff
@@ -101,24 +115,38 @@ impl Manifest {
...
+                Some(entry) => vec![entry.variant.as_deref()],
+                None => entries
+                    .iter()
+                    .filter(|entry| entry.path.starts_with(&path) && entry.path != path)
+                    .map(|entry| entry.variant.as_deref())
+                    .collect::<BTreeSet<_>>()
+                    .into_iter()
+                    .collect(),
+            };
+            for variant in variants {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/manifest.rs]: Confirmed. f7ee0087b records a containing directory once, under the variant-less path (it is one filesystem object), and the pull planner keeps one step per resolved directory, an enrolled directory own stream outranking the containing record. Tests: a_containing_directory_is_recorded_once_across_streams and one_step_per_directory_across_claims.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/manifest.rs [hunk 4]
```diff
@@ -70,10 +73,21 @@ impl Manifest {
...
+        };
+        match self.owner(&portable) {
+            Some(entry) => in_stream(entry),
+            None => {
+                !relative.is_empty()
+                    && self
+                        .enrollment
+                        .iter()
+                        .filter(|entry| strictly_below(&entry.path, &portable))
+                        .any(in_stream)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/manifest.rs]: Confirmed. f7ee0087b makes the containing-directory rule apply whenever the enclosing owner does not match the stream (not only when there is no owner), so the inactive macOS directory enrollment no longer hides the Linux parent record. Test: an_inactive_directory_enrollment_keeps_the_parent_record.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/checkpoint.rs [hunk 5]
```diff
@@ -736,7 +750,11 @@ impl Store {
                     .tree
                     .modes
                     .into_iter()
-                    .filter(|(path, _)| entries.iter().any(|entry| under_entry(path, entry)))
+                    .filter(|(path, _)| {
+                        entries
+                            .iter()
+                            .any(|entry| under_entry(path, entry) || under_entry(entry, path))
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/checkpoint.rs:761]: Same as the Bugbot thread on this line: f7ee0087b fixes read_meta so containing directories reach tree.modes at all, and adds the observed-directory guard to the selective promotion path, with a regression test that fails without it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 6]
```diff
@@ -45,30 +77,47 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
-        if !super::run::eligible(&roots, tracked, portable)
-            || repo
-                .object_at(tree, portable)?
-                .is_none_or(|(mode, _)| mode != "040000")
+        let Some(claim) = claim(&roots, tracked, portable) else {
+            continue;
+        };
+        if repo
+            .object_at(tree, portable)?
+            .is_none_or(|(mode, _)| mode != "040000")
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed. 8f136b9b0 accepts a containing record when a tree exists at its own path or at the path of any stream selected here for an enrolled path inside it (via the entries already-selected variants, no hard-coded os), so a Linux-only file still gets its private parent on a fresh machine. Unit test a_parent_present_only_in_a_variant_stream_is_planned (fails without the fix) and the e2e now covers a --os entry end to end.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 7]
```diff
@@ -45,30 +77,47 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
+        }
+    }
+    for (path, (_, portable)) in claimed {
         let before = observe(&path)?;
         let desired = tracked
             .manifest
             .permissions
             .get(portable)
             .copied()
             .unwrap_or(0o755);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Reachable, but the fallback is the intended result rather than a bug: incoming_tracking merges base/local/remote, so a per-stream key is absent from the incoming manifest only when a machine with that stream active saved the directory at the default, and the enrolled stream (including its absence) is authoritative for the machine that owns it, which is why Enrolled outranks Containing. 8f136b9b0 pins this with a test stage in one_step_per_directory_across_claims (local head still records the own stream, incoming dropped it, incoming containing record 0700, expected desired 0755).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 8]
```diff
@@ -45,30 +77,47 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
-                .object_at(tree, portable)?
-                .is_none_or(|(mode, _)| mode != "040000")
+        let Some(claim) = claim(&roots, tracked, portable) else {
+            continue;
+        };
+        if repo
+            .object_at(tree, portable)?
+            .is_none_or(|(mode, _)| mode != "040000")
         {
             continue;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed (same mechanism as the Bugbot thread on this line). 8f136b9b0 lets the planner find the directory under any stream selected here for an enrolled path inside it, so the variant-only parent is recreated 0700; unit test a_parent_present_only_in_a_variant_stream_is_planned plus an --os entry in the e2e.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/manifest.rs [hunk 9]
```diff
@@ -101,26 +113,39 @@ impl Manifest {
...
+        // an active entry replaces its own path, everything below it, and
+        // (in the variant-less stream) the directories between it and the root
         self.permissions.retain(|path, _| {
             !active.iter().any(|prefix| {
                 path == prefix
-                    || path
-                        .strip_prefix(prefix)
-                        .is_some_and(|rest| rest.starts_with('/'))
+                    || strictly_below(path, prefix)
+                    || (!path.contains('@') && strictly_below(&plain(prefix), path))
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/manifest.rs]: Confirmed: the literal @ check matched a directory name. 8f136b9b0 detects a plain path as one whose stream marker parse is the identity (plain(path) == path, the same first-component parse owns_stream uses), so a 0700 ~/.private@work returns to default on save; test a_directory_named_with_an_at_sign_returns_to_default fails without the fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 10]
```diff
@@ -45,30 +77,73 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
+        for candidate in &candidates {
+            if repo
+                .object_at(tree, candidate)?
+                .is_some_and(|(mode, _)| mode == "040000")
+            {
+                present = true;
+                break;
+            }
+        }
+        if !present {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed: was_directory only looked the containing record up at the variant-less path in the local tree. 1b3362bcf keeps every key that claims a directory (best first) and treats it as previously saved when it exists in the local tree under any of those keys or their stream paths, reading the saved bits from the first key with a record; the variant-stream test now also asserts an unsaved 0711 chmod is reported instead of planned over, and fails without the fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 11]
```diff
@@ -45,53 +85,108 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
+                {
+                    was_directory = true;
+                    break;
+                }
+            }
+        }
+        let saved_bits = keys
+            .iter()
+            .find_map(|key| saved.permissions.get(key.portable).copied())
+            .unwrap_or(0o755);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed reachable (local: enrolled key dropped and 0755 applied, containing 0700 kept; incoming: enrolled key 0750). 290099891 reads the local record under the best key whose stream the local head already has and treats its absence as 0755, deferring to the containing record only when that stream is new locally, so the chmod is planned with no false conflict and a real 0711 chmod is still reported; test a_dropped_enrolled_record_means_the_default_locally fails without the rule.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 12]
```diff
@@ -45,53 +85,116 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
+                    if repo
+                        .object_at(head, candidate)?
+                        .is_some_and(|(mode, _)| mode == "040000")
+                    {
+                        saved_bits = Some(
+                            saved
+                                .permissions
+                                .get(key.portable)
+                                .copied()
+                                .unwrap_or(0o755),
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed, the mirror of the previous case: tree presence was read as evidence of enrollment. 57d437bb0 states the invariant once and implements it from the saved local manifest enrollment alone: (a) the record under the key whose stream that enrollment governs for the directory itself, absence meaning 0755; (b) else the local containing record if the manifest has one; (c) else 0755 if the enrollment knows the directory at all; otherwise no baseline. Test a_newly_enrolled_parent_keeps_its_containing_baseline covers this exact sequence (fails with the old rule), and the dropped-record, superseded-key, variant-stream and no-local-record tests still pass.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 13]
```diff
@@ -45,53 +83,119 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
+        let containing = roots.branch_path(&path, None);
+        let saved_bits = keys
+            .iter()
+            .find(|key| saved.enrolls_stream(key.portable))
+            .map(|key| {
+                saved
+                    .permissions
+                    .get(key.portable)
+                    .copied()
+                    .unwrap_or(0o755)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed by construction. 551b00528 applies the guard only to a live directory: when observe() is None there is no mode to protect, so the directory is created with the incoming mode; test a_new_private_subdirectory_is_created_not_guarded reproduces the sequence and reports the exact "unsaved" bail without the fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 14]
```diff
@@ -45,53 +83,119 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
+            .or_else(|| {
+                containing
+                    .as_ref()
+                    .and_then(|containing| saved.permissions.get(containing).copied())
+            })
+            .or_else(|| {
+                containing
+                    .as_ref()
+                    .is_some_and(|containing| saved.owns_stream(containing))
+                    .then_some(0o755)
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed. 551b00528 derives the baseline from the saved enrollment ownership of the directory under the variant this machine selects (Manifest::governing_key), absence meaning 0755, so a directory reduced from its own variant enrollment to a containing record keeps its record; test a_removed_variant_enrollment_keeps_its_baseline is the inverse of the newly-enrolled case and is overwritten without the fix.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 15]
```diff
@@ -45,53 +83,127 @@ pub(super) fn plan(repo: &HistoryRepo, tracked: &TrackedSet, tree: &str) -> Resu
...
                 desired,
                 written: None,
             });
+            continue;
+        };
+        let containing = roots.branch_path(&path, None);
+        let saved_bits = containing
+            .as_ref()
+            .and_then(|containing| saved.governing_key(containing))
+            .map(|key| saved.permissions.get(&key).copied().unwrap_or(0o755))
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed as the root fault: the mode a pull applied and the baseline the next pull assumed came from two rules. 0550b757d makes them one computation (mode_from: the enrolled stream record covering the directory, absent meaning the default, else the containing record, absent meaning the default), run against the incoming manifest for what to apply and against the saved manifest, with this machine variant selection, for the baseline; the guard is now "live mode differs from what this planner would have applied from the saved manifest", and no live directory means no guard. Test adoption_and_the_next_pull_agree_on_the_baseline covers the sequence, and one_step_per_directory_across_claims stage 2 now expects 0755 (it encoded the old inconsistency).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/manifest.rs [hunk 16]
```diff
@@ -60,17 +60,51 @@ impl Manifest {
...
+
+    /// The permission key under which this enrollment governs a directory
+    /// on this machine: the directory (or an enrolled path above it) is
+    /// enrolled, under its selected variant if it has any. `None` when the
+    /// enrollment does not govern the directory itself here, whether because
+    /// only paths inside it are enrolled or because no variant is selected.
+    pub(crate) fn governing_key(&self, path: &str) -> Option<String> {
+        let entry = self.owner(path)?;
+        let environments = super::select::active_environments();
+        match super::select::select(&entry.variants, &environments) {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/manifest.rs]: Confirmed, and folded into the unified baseline in 0550b757d rather than fixed separately: the baseline is computed from the enrollment this machine selects from the saved manifest (Manifest::selected_entries skips an enrollment with no selected variant), so a nested Windows-only enrollment no longer hides the enclosing enrollment on Linux; validation likewise accepts a stream that any enclosing enrollment exposes. Test an_inactive_nested_enrollment_does_not_hide_the_governing_parent covers the exact sequence.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 17]
```diff
@@ -34,6 +38,96 @@ fn observe(_path: &std::path::Path) -> Result<Option<(u64, u64, u32)>> {
...
+            && let Ok(stream) = entry.tree_path(path)
+            && !candidates.contains(&stream)
+        {
+            candidates.push(stream);
+        }
+    }
+    for candidate in &candidates {
+        if repo
+            .object_at(tree, candidate)?
+            .is_some_and(|(mode, _)| mode == "040000")
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs]: Confirmed. d5c1e1955 decides presence in the governing stream: for a path an enrolled path covers, only the object at its governing key counts, and a file or link there means no directory step regardless of what other streams hold; a containing directory needs a directory in some inner stream and nothing else at that path in any of them. Test a_file_in_the_governing_stream_is_not_made_a_directory reproduces the mixed-stream bootstrap and plans the bogus 0600 directory step under the old semantics.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/manifest.rs [hunk 18]
```diff
@@ -60,20 +60,73 @@ impl Manifest {
...
+        let (_, variant) = stem
+            .split_once('@')
+            .map_or((stem, None), |(root, variant)| (root, Some(variant)));
+        let portable = plain(path);
+        self.enrollment
+            .iter()
+            .filter(|entry| portable == entry.path || strictly_below(&portable, &entry.path))
+            .any(|entry| match variant {
+                Some(name) => entry.variants.iter().any(|variant| variant.name() == name),
+                None => entry.variants.is_empty(),
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/manifest.rs:128]: Confirmed. d5c1e1955 rebuilds a checkpoint directory mode only from the record under the key that governs the path on this machine (the same governing_key the planner uses, closest selected enrollment first), so another platform containing record for a directory whose own stream records the default no longer lands in this machine checkpoint or its rollback. Test read_meta_takes_a_directory_mode_from_its_governing_stream_only covers the exact case (no 0700 for private with the nested enrollment; 0700 without it) and fails under the any-enclosing-entry check.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: src/system/history/sync/directories.rs [hunk 19]
```diff
@@ -34,6 +38,83 @@ fn observe(_path: &std::path::Path) -> Result<Option<(u64, u64, u32)>> {
...
+            .object_at(tree, candidate)?
+            .map(|(mode, _)| mode == "040000"))
+    };
+    let Some(governing) = governing_key(roots, entries, path) else {
+        return Ok(false);
+    };
+    let covered = entries.iter().any(|entry| path.starts_with(&entry.path));
+    if covered {
+        return Ok(is_directory(&governing)? == Some(true));
+    }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/system/history/sync/directories.rs:97]: Confirmed as the over-correction. 309250b51 makes presence three-way for a covered path: a file or link in the governing stream means no directory step, a directory there means present, and nothing there falls through to the containing rule (a directory in some inner stream and nothing else at that path in any stream). Test a_covered_directory_present_only_in_an_inner_stream_is_planned reproduces the Linux-only-settings case (0 steps under the old rule, the 0700 step now) and a_file_in_the_governing_stream_is_not_made_a_directory still passes.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
